server/middleware/repo_auth.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2018 Palantir Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package middleware
+
+import (
+	"net/http"
+	"strings"
+
+	"github.com/bluekeyes/hatpear"
+	"github.com/palantir/go-githubapp/githubapp"
+	"github.com/pkg/errors"
+	"goji.io/pat"
+)
+
+type TokenValidator interface {
+	// ValidateRepoAccess checks that the passed in token is valid and that it can read the repo.
+	ValidateRepoAccess(r *http.Request, token string) (valid bool, err error)
+}
+
+// RepoAuth returns middleware that rejects requests if they do not include an `owner` and `repo`
+// param and a valid GitHub token with access to the provided repo. Supports both personal and
+// server to server tokens.
+func RepoAuth(validator TokenValidator) func(http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return hatpear.TryFunc(func(w http.ResponseWriter, r *http.Request) error {
+			ctx := r.Context()
+
+			token := getBearerToken(r)
+			if token == "" {
+				return errors.New("missing token")
```

Comment:
All of these will turn in to generic "500 Internal Server Error" responses, which is not very helpful for authentication issues. I think we'll want to either write a response directly from this middleware or define a custom error that implements the [`StatusCode()`](https://github.com/palantir/go-baseapp/blob/develop/baseapp/error.go#L29-L31) method to take advantage of baseapp's ability to change response codes.